File: docs/DSA/src/go/lru_cache.go
package main

type DoubleListNode struct {
	Key  int
	Val  int
	Next *DoubleListNode
	Prev *DoubleListNode
}

type LRUCache struct {
	Cap  int
	data map[int]*DoubleListNode
	MRU  *DoubleListNode
	LRU  *DoubleListNode
}

func ConstructorLRU(capacity int) LRUCache {
	obj := LRUCache{Cap: capacity,
		data: make(map[int]*DoubleListNode),
		MRU:  &DoubleListNode{Key: -1, Val: -1},
		LRU:  &DoubleListNode{Key: -1, Val: -1},
	}
	obj.LRU.Next, obj.MRU.Prev = obj.MRU, obj.LRU
	return obj
}

func (this *LRUCache) remove(node *DoubleListNode) {
	prev, next := node.Prev, node.Next
	prev.Next, next.Prev = next, prev
}

func (this *LRUCache) insert(node *DoubleListNode) {
	prev, nxt := this.MRU.Prev, this.MRU
	prev.Next = node
	nxt.Prev = node
	node.Prev, node.Next = prev, nxt
}

func (this *LRUCache) Get(key int) int {
	if node, ok := this.data[key]; ok {
		this.remove(node)
		this.insert(node)
		return node.Val
	}
	return -1
}

func (this *LRUCache) Put(key int, value int) {
	if node, ok := this.data[key]; ok {
		this.remove(node)
	}
	node := &DoubleListNode{Key: key, Val: value}
	this.data[key] = node
	this.insert(node)
	if len(this.data) > this.Cap {
		lru := this.LRU.Next
		this.remove(lru)
		delete(this.data, lru.Key)
	}
}

File: DSA/src/py/lru_cache.py
class ListNode:
    def __init__(self, key=0, val=0):
        self.key, self.val = key, val
        self.prev = self.next = None

class LRUCache:

    def __init__(self, capacity: int):
        self.cap = capacity
        self.data = dict()

        self.lru, self.mru = ListNode(), ListNode()
        self.lru.next, self.mru.prev = self.mru, self.lru

    def remove(self, node):
        prev, nxt = node.prev, node.next
        prev.next, nxt.prev = nxt, prev

    def insert(self, node):
        prev, nxt = self.mru.prev, self.mru
        prev.next = nxt.prev = node
        node.prev, node.next = prev, nxt

    def get(self, key: int) -> int:
        if key in self.data:
            self.remove(self.data[key])
            self.insert(self.data[key])
            return self.data[key].val
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.data:
            self.remove(self.data[key])

        self.data[key] = ListNode(key=key, val=value)
        self.insert(self.data[key])

        if len(self.data) > self.cap:
            lru = self.lru.next
            self.remove(lru)
            del self.data[lru.key]

